Long-running daemon

thallgren opened the pull request:
## Description

This PR refactors the code of the root daemon and changes its behavior so that it doesn't terminate when a connection ends.  Internally, the connection (and the TUN-device and DNS-resolver) is maintained by a session which has the same lifecycle as the connection.

Three significant changes deserves a mention:
1. The daemon.log is rotated daily instead of once per connection.
2. The config.yml is loaded automatically when it is modified.
3. The messages printed when starting and quitting a connection mentions "network" rather than "root daemon".

## Checklist

<!--
  Please review the requirements for each checkbox, and check them
  off (change "[ ]" to "[x]") as you verify that they are complete.
-->

 - [ ] I made sure to update `./CHANGELOG.md`.
 - [ ] I made sure to add any docs changes required for my change (including release notes).
 - [x] My change is adequately tested.

Commit:
Remove obsolete and no longer used http client in the root daemon

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Prevent segfault when obtaining default config and Env is unset

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Rename outbound struct to session

The outbound struct really represents what the daemon will need to do
in order to manage one "connect" session from the client.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Merge session (former outbound) and tunrouter

The split between 1:1 mapped session and router didn't
make much sense.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Rename dns/dns.go to dns/server.go

Because it contains the dns.Server

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Don't set and report log level unless it actually changes

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Refactor the root daemon code

* Moves all dns related things into the dns-package
* Moves all TUN-device routing in to a router.go file
* Ensures creating of an immutable session by dividing initialization
  into the stages:
  1. Connect to the traffic-manager and verify its version
  2. Start watching traffic-manager's cluster info.
  3. Configure TUN-device and start DNS upon receipt of first cluster
     info

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Don't pass dnsIP as a root daemon command line argument.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Rename and change some root daemon methods

Renames the root daemon's `SetOutboundInfo` to `Connect` and adds a
`Disconnect` which disconnects the session but doesn't stop the daemon.
The `Quit` will disconnect and stop the daemon.

Renames methods in the cli package.
- DidLaunchDaemon -> DidLaunchNetwork
- QuitDaemon -> Disconnect
- WithDaemon -> WithNetwork
- WithStartedDaemon -> WithStartedNetwork

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add auto-reload of the config.yml to the root daemon

Adds a [fsnotify](https://github.com/fsnotify/fsnotify) watcher that
ensures that the `config.yml` file is reloaded when updated. The
reload affects the `Config` returned by `client.GetConfig` atomically so
that a `Config` instance is returned on the next call.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix tests

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Don't let CLI explicitly quit connector. Perform a disconnect instead

The disconnect terminates the daemon session and quits the connector.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Ensure proper close of UserNotification stream on quit

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Rotate root daemon log once a day

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

josecv commented on pkg/client/daemon/service.go:
```diff
@@ -49,6 +50,9 @@ type service struct {
 	cancel        context.CancelFunc
+	sessionsCtx   context.Context
+	sessionLock   sync.Mutex
+	session       *session
```
is this a dupe? there's another `session *session` on line 51

thallgren replied:
Not sure what's going on here. Must be a glitch in how the diff is displayed somehow. There's only one session in the struct and that's the one below the `sessionsLock`.

josecv commented on pkg/client/daemon/service.go:
```diff
@@ -197,38 +233,15 @@ func run(c context.Context, loggingDir, configDir, dns string) error {
-	g.Go("server-router", func(ctx context.Context) error {
-		defer scoutUsers.Done()
-		return d.session.run(ctx)
+	g.Go("session", func(c context.Context) (err error) {
+		c, d.cancel = context.WithCancel(c)
+		d.sessionsCtx = c
+		<-c.Done()
+		return nil
```
hmm i feel like the handling of the sessions' lifecycle is a little bit awkward in general. I don't really like the context being stored in the struct here, or that successive sessions share the same context.
It's also a little hard to tell by looking at `daemon/service.go` where exactly the session's goroutines are getting started.

IMO what we should consider doing is split `newSession` into a `newSession` that doesn't start any new goroutines and a `runSession` that actually kicks everything off.
We could then have this goroutine (`g.Go("session")`) receive the `newSession` on some channel and `runSession` on it. Here's sort of what I'm picturing:

```
type service struct {
	sessionCh      chan *session
        sessionErrCh chan error
	sessionLock   sync.Mutex
	session           *session
}

func (d *service) SetOutboundInfo(ctx context.Context, info *rpc.OutboundInfo) (*empty.Empty, error) {
	d.sessionLock.Lock()
	defer d.sessionLock.Unlock()
	if d.session != nil {
		return nil, status.Error(codes.AlreadyExists, "an active session exists")
	}
	s, err := newSession(ctx, d.scout, info)
	if err != nil {
		return nil, status.Error(codes.Internal, err.Error())
	}
	d.sessionCh <- s
        err = <-d.sessionErrCh
        // handle err
        d.session = s // session was started successfully, so store it
	return &empty.Empty{}, nil
}

g.Go("session", func(c context.Context) (err error) {
      c, d.cancel = context.WithCancel(c)
      for {
          sess := <-d.sessionCh
          sCtx, sCancel := context.WithCancel(c)
          doneCh, err := sess.runSession(sCtx)
          d.sessionErrCh <- err
          select {
                case <-c.Done():
                     sCancel()
                     return nil
                case <-doneCh: // assuming doneCh is already closed if runSession failed
                     sCancel()
          }
      }
})
```

This looks a bit bulkier than what you have but i think it gives us a few advantages:

* opens it up to refactor session cancellation so that just doing `sCancel()` is good enough to stop the session -- eventually we could just store `sCancel` and call it in `disconnect`, dropping the need for `session.stop()`. This would be good because right now it's not clear to me that `d.cancel` will clean everything up correctly
* makes it really clear that the sessions are running in the "session" goroutine
* prevents the sessions' context from being shared by multiple sessions, or leaked to the rest of this struct

thallgren replied:
Not sure I understand your reasoning here. The correspondence to your `sCancel` is currently stored as `cancel` in the `session` struct. It's created by the `newSession` function from the parent `ctx` which is the `d.sessionsCtx`.  A `disconnect` calls `d.session.stop` to do necessary clean-up. That method in turn calls that `cancel` and terminates the session.

How would `disconnect` get hold of the `sCancel`? It needs to be stored somewhere. What place is better than the session that it cancels?

Please note that the `ctx` received by `Connect` is short lived. Using that context would terminate the connection as soon as the connect call returns.  Hence the need for the `d.sessionsCtx` which lives until the process terminates.

thallgren replied:
`Connect` in my reasoning is `SetOutboundInfo` before it was renamed.

thallgren replied:
Your design is cleaner so I decided to go ahead and implement something similar. The new design is based on your channel pair and moving the actual session execution into the "session" goroutine. It differs slightly in how the cancellation is handled but there's no explicit call to `stop` now. The session can only be cancelled by calling its `cancel` (which is the context cancellation method).

josecv reviewed (changes requested):
this all looks good but it'd be good to discuss about session lifecycles as per my other comment

Commit:
More distinct session lifecycle logic

Adds the two channels connectCh and connectErr to the service and moves
the creation and run of the session into the "session" go-routine.

This ensures that session cancellation is consistent regardless of which
of the parent contexts it is that get cancelled.

Thanks @josecv for the ideas on how to improve this.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Reduce the cognitive complexity of the service run method

No change in logic, just move inlined methods out of the run method.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Only use Suite.Require from within tests.

Extracting the `Require` interface from a parent test is bad and results
in panics if the test fails.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Ensure that links and routes are restored

Relying on context for cancellation made some paths more
sensitive. This commit ensures that routes and links are
restored, even when the context is cancelled.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>
